pkg/operation/cloudbotanist/packetbotanist/controlplane.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package packetbotanist
+
+import (
+	"github.com/gardener/gardener/pkg/operation/common"
+)
+
+// GenerateCloudProviderConfig generates the Packet cloud provider config.
+// See this for more details:
+// https://github.com/packethost/packet-ccm
+func (b *PacketBotanist) GenerateCloudProviderConfig() (string, error) {
+	return "", nil
+}
+
+// RefreshCloudProviderConfig refreshes the cloud provider credentials in the existing cloud
+// provider config.
+// Not needed on Packet (cloud provider config does not contain the credentials), hence, the
+// original is returned back.
+func (b *PacketBotanist) RefreshCloudProviderConfig(currentConfig map[string]string) map[string]string {
+	return currentConfig
+}
+
+// GenerateKubeAPIServerServiceConfig generates the cloud provider specific values which are required to render the
+// Service manifest of the kube-apiserver-service properly.
+func (b *PacketBotanist) GenerateKubeAPIServerServiceConfig() (map[string]interface{}, error) {
+	return map[string]interface{}{}, nil
+}
+
+// GenerateKubeAPIServerExposeConfig defines the cloud provider specific values which configure how the kube-apiserver
+// is exposed to the public.
+func (b *PacketBotanist) GenerateKubeAPIServerExposeConfig() (map[string]interface{}, error) {
+	return map[string]interface{}{}, nil
+}
+
+// GenerateKubeAPIServerConfig generates the cloud provider specific values which are required to render the
+// Deployment manifest of the kube-apiserver properly.
+func (b *PacketBotanist) GenerateKubeAPIServerConfig() (map[string]interface{}, error) {
+	return map[string]interface{}{
+		"environment": getPacketCredentialsEnvironment(),
+	}, nil
+}
+
+// GenerateCloudControllerManagerConfig generates the cloud provider specific values which are required to
+// render the Deployment manifest of the cloud-controller-manager properly.
+func (b *PacketBotanist) GenerateCloudControllerManagerConfig() (map[string]interface{}, string, error) {
+	return map[string]interface{}{
```

Comment:
The Alicloud `GenerateCloudControllerManagerConfig()` looks as follows:

```go
func (b *AlicloudBotanist) GenerateCloudControllerManagerConfig() (map[string]interface{}, string, error) {
        chartName := "alicloud-cloud-controller-manager"
        conf := map[string]interface{}{
                "configureRoutes": false,
        }
        newConf, err := b.ImageVector.InjectImages(conf, b.SeedVersion(), b.ShootVersion(), common.AlicloudControllerManagerImageName)
        if err != nil {
                return conf, chartName, err
        }

        return newConf, chartName, nil
}
```

Does the above just populate the chart at `charts/seed-controlplane/charts/alicloud-cloud-controller-manager/`?

It would help if I knew what is expected to be returned by `GenerateCloudControllerManagerConfig()`. What is supposed to be in the `map[string]interface{}` and `string`? And what does it do with those results?

Reply:
Before Alicloud support got added Gardener used the `hyperkube` image built by the Kubernetes commnunity (which contains a cloud-controller-manager binary for AWS, Azure, GCP, and OpenStack). The problem with Alicloud was that this `hyperkube` image didn't support the CCM for Alicloud - that's why @jia-jerry created a seperate chart (see https://github.com/gardener/gardener/tree/master/charts/seed-controlplane/charts/alicloud-cloud-controller-manager) for the Alicloud CCM (which uses [another image](https://github.com/gardener/gardener/blob/master/charts/images.yaml#L102-L106) and other configuration options).

During the shoot reconciliation flow Gardener calls the [`DeployCloudControllerManager`](https://github.com/gardener/gardener/blob/master/pkg/operation/hybridbotanist/controlplane.go#L506-L550) function. As usual, this function computes the Helm chart values for the CCM chart (https://github.com/gardener/gardener/tree/master/charts/seed-controlplane/charts/cloud-controller-manager) and deploys it. Now Alicloud is different/special (it has its own chart and image, however, most of the configuration options are similar). [Here](https://github.com/gardener/gardener/blob/master/pkg/operation/hybridbotanist/controlplane.go#L525) the cloud botanist is called (which, for Alicloud, is the function you pasted above). It injects values and overwrites the Docker image (not `hyperkube` but the Alicloud-specific one).
Also, it returns the name of the chart containing the Alicloud CCM: https://github.com/gardener/gardener/tree/master/charts/seed-controlplane/charts/alicloud-cloud-controller-manager

You probably want to do the same: Create a chart for the Packet CCM, overwrite the image (`hyperkube` does not contain the packet CCM, similar to Alicloud) and compute the needed values for this chart.

Please note that all of this will get much easier/cleaner with extensibility. The provider-specific controller will be completely responsible for configuration and deployment of its cloud-controller-manager (and also CSI controllers, etc.), see e.g. https://github.com/gardener/gardener-extensions/tree/master/controllers/provider-aws/pkg/controller/controlplane (it's not yet completely finished, but you can get the idea).

Reply:
I think I get it. Will give it a shot.

Reply:
I _think_ I got what you are suggesting. Adding the appropriate chart in `seed-controlplane/charts/packet-cloud-controller-manager/` and updated `GenerateCloudControllerManagerConfig()` here.

One open question: the CCM needs the API key and project ID to do its work. In the chart, I left it as our standard template:

```yml
        env:
          - name: PACKET_AUTH_TOKEN
            valueFrom:
              secretKeyRef:
                name: packet-cloud-config
                key: apiKey
          - name: PACKET_PROJECT_ID
            valueFrom:
              secretKeyRef:
                name: packet-cloud-config
                key: projectID
```

But that cannot be right since the secret `packet-cloud-config` doesn't exist.

How do we get that injected?

Reply:
> How do we get that injected?

You already let the end-user provide this information: https://github.com/gardener/gardener/blob/master/example/70-secret-cloudprovider-packet.yaml#L12-L13. Gardener copies this information into a `Secret` in the seed named `cloudprovider`.

Reply:
And in the same namespace, so I can just do:

```yml
        env:
          - name: PACKET_AUTH_TOKEN
            valueFrom:
              secretKeyRef:
                name: cloudprovider
                key: apiToken
          - name: PACKET_PROJECT_ID
            valueFrom:
              secretKeyRef:
                name: cloudprovider
                key: projectID
```

Correct? Updating now.

Reply:
Updated.